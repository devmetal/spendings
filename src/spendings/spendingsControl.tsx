import React, { FC } from 'react';
import Row from 'react-bootstrap/Row';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Container';
import { TSpending } from '../types';
import { SpendingInput } from './spendingInput';
import { Spendings } from './spendings';

export interface SpendingsControlProps {
  spendings: Array<TSpending>;
  onDelete: (id: number) => void;
  onUpdate: (spending: TSpending) => void;
  onCreate: (spending: { amount: number; name: string }) => void;
}

export const SpendingsControl: FC<SpendingsControlProps> = ({
  spendings,
  onCreate,
  onDelete,
  onUpdate,
}) => {
  return (
    <Container>
      <Row>
        <Col>
          <SpendingInput onSave={onCreate} disabled={false} />
        </Col>
      </Row>
      <Row>
        <Col>
          <Spendings
            spendings={spendings}
            onDelete={onDelete}
            onUpdate={onUpdate}
          />
        </Col>
      </Row>
    </Container>
  );
};
