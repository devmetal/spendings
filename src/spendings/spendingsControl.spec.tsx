import { fireEvent, render, screen, wait } from '@testing-library/react';
import React from 'react';
import { SpendingsControl } from './spendingsControl';

const props = {
  onCreate: jest.fn(),
  onUpdate: jest.fn(),
  onDelete: jest.fn(),
  spendings: [
    { id: 1, amount: 100, name: 'Teszt' },
    { id: 2, amount: 101, name: 'Teszt1' },
    { id: 3, amount: 102, name: 'Teszt2' },
  ],
};

test('Sanity snapshot test', () => {
  const { container } = render(<SpendingsControl {...props} />);
  expect(container).toMatchSnapshot();
});

describe('Interaction test', () => {
  test('update a spending', async () => {
    const { getByText, queryByText, getAllByText, getAllByTestId } = render(
      <SpendingsControl {...props} />
    );
    fireEvent.click(getByText('Teszt'));

    /**
     * Wait for form is visible
     */
    await wait(() => expect(getByText('Cancel')).toBeVisible());

    fireEvent.change(getAllByTestId('name')[1], {
      target: { value: 'Updated' },
    });
    fireEvent.change(getAllByTestId('amount')[1], {
      target: { value: '500' },
    });
    fireEvent.click(getAllByText('Save')[1]);

    /**
     * Form has to be dissapear
     */
    await wait(() => expect(queryByText('Cancel')).toBeNull());

    expect(props.onUpdate).toBeCalledWith({
      id: 1,
      amount: 500,
      name: 'Updated',
    });
  });
});
