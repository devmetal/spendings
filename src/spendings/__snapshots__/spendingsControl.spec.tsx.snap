// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity snapshot test 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="container"
      >
        <div
          class="mb-3 input-group"
        >
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Name
            </span>
          </div>
          <input
            class="form-control"
            data-testid="name"
            type="text"
            value=""
          />
          <div
            class="input-group-prepend"
          >
            <span
              class="input-group-text"
            >
              Amount
            </span>
          </div>
          <input
            class="form-control"
            data-testid="amount"
            min="1"
            type="number"
            value=""
          />
          <div
            class="input-group-prepend"
          >
            <button
              class="btn btn-outline-primary"
              type="button"
            >
              Save
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="container"
      >
        <div
          class="list-group"
        >
          <div
            class="list-group-item"
          >
            <div
              class="list-group list-group-horizontal"
            >
              <div
                class="list-group-item"
              >
                Teszt
              </div>
              <div
                class="list-group-item"
              >
                100
              </div>
              <div
                class="list-group-item list-group-item-danger"
              >
                Delete
              </div>
            </div>
          </div>
          <div
            class="list-group-item"
          >
            <div
              class="list-group list-group-horizontal"
            >
              <div
                class="list-group-item"
              >
                Teszt1
              </div>
              <div
                class="list-group-item"
              >
                101
              </div>
              <div
                class="list-group-item list-group-item-danger"
              >
                Delete
              </div>
            </div>
          </div>
          <div
            class="list-group-item"
          >
            <div
              class="list-group list-group-horizontal"
            >
              <div
                class="list-group-item"
              >
                Teszt2
              </div>
              <div
                class="list-group-item"
              >
                102
              </div>
              <div
                class="list-group-item list-group-item-danger"
              >
                Delete
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
